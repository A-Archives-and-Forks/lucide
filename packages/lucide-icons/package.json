{
  "name": "@lucide/icons",
  "description": "A Lucide icon library that contains icon data in a standard Lucide format.",
  "version": "0.0.1",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide"
  },
  "keywords": [
    "Lucide",
    "HTML",
    "Feather",
    "Icons",
    "Icon",
    "SVG",
    "Feather Icons"
  ],
  "amdName": "lucide-icons",
  "source": "src/lucide-icons.ts",
  "main": "dist/cjs/lucide-icons.cjs",
  "module": "dist/esm/lucide-icons.mjs",
  "types": "dist/esm/lucide-icons.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/esm/lucide-icons.d.ts",
      "import": "./dist/esm/lucide-icons.mjs",
      "require": "./dist/cjs/lucide-icons.cjs",
      "default": "./dist/esm/lucide-icons.mjs"
    },
    "./icons/*": {
      "import": "./dist/esm/icons/*.mjs",
      "default": "./dist/esm/icons/*.mjs"
    },
    "./dynamic": {
      "types": "./dist/esm/dynamic.d.ts",
      "import": "./dist/esm/dynamic.mjs",
      "require": "./dist/cjs/dynamic.cjs",
      "default": "./dist/esm/dynamic.mjs"
    },
    "./build": {
      "types": "./dist/esm/build.d.ts",
      "import": "./dist/esm/build.mjs",
      "require": "./dist/cjs/build.cjs",
      "default": "./dist/esm/build.mjs"
    }
  },
  "scripts": {
    "build": "pnpm clean && pnpm copy:license && pnpm build:icons && pnpm build:bundle",
    "copy:license": "cp ../../LICENSE ./LICENSE",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.ts",
    "build:icons": "build-icons --output=./src --templateSrc=./scripts/exportTemplate.mjs --renderUniqueKey --withAliases --withDynamicImports --separateAliasesFile --aliasesFileExtension=.ts --iconFileExtension=.ts --exportFileName=index.ts",
    "build:bundle": "rollup -c rollup.config.mjs",
    "test": "pnpm build:icons && vitest run",
    "test:watch": "vitest watch",
    "test:update": "vitest -u",
    "version": "pnpm version --git-tag-version=false"
  },
  "devDependencies": {
    "@lucide/build-icons": "workspace:*",
    "@lucide/helpers": "^1.0.0",
    "@lucide/rollup-plugins": "workspace:*",
    "@rollup/plugin-replace": "^6.0.2",
    "@testing-library/jest-dom": "^6.6.3",
    "jest-serializer-html": "^7.1.0",
    "rollup": "^4.40.0",
    "rollup-plugin-dts": "^6.2.1",
    "rollup-plugin-preserve-directives": "^0.4.0",
    "typescript": "^4.9.5",
    "vite": "5.4.14",
    "vitest": "^1.6.1"
  }
}
