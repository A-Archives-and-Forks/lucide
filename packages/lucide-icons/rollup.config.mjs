import plugins from '@lucide/rollup-plugins';
import pkg from './package.json' with {type: 'json'};
import dts from 'rollup-plugin-dts';

const packageName = '@lucide/icons';
const outputFileName = 'lucide-icons';
const inputs = [`src/lucide-icons.ts`];
const bundles = [
  {
    format: 'umd',
    inputs,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
  },
  {
    format: 'cjs',
    inputs,
    extension: 'cjs',
  },
  {
    format: 'esm',
    inputs: [...inputs, './src/dynamic.ts', './src/build.ts'],
    preserveModules: true,
    extension: 'mjs',
  },
  /*
  {
    format: 'esm',
    inputs: ['src/dynamic.ts'],
    outputFileName: 'dynamic',
    external: [/src/],
    paths: (id) => {
      if (id.match(/src/)) {
        const [, modulePath] = id.match(/src\/(.*)\.ts/);

        return `${modulePath}.js`;
      }
    },
  },
  {
    format: 'esm',
    inputs: ['src/build.ts'],
    outputFileName: 'build.mjs',
    external: [/src/],
    paths: (id) => {
      if (id.match(/src/)) {
        const [, modulePath] = id.match(/src\/(.*)\.ts/);

        return `${modulePath}.js`;
      }
    },
  },
   */
];

const configs = bundles
  .map(
    ({
       inputs,
       outputDir = 'dist',
       outputFile,
       format,
       minify,
       preserveModules,
       entryFileNames,
       external = [],
       paths,
       extension = 'js',
     }) =>
      inputs.map((input) => ({
        input,
        plugins: [
          ...plugins({pkg, minify}),
        ],
        external,
        output: {
          name: packageName,
          entryFileNames,
          ...(preserveModules
            ? {
              dir: `${outputDir}/${format}`,
              entryFileNames: `[name].${extension}`,
            }
            : {
              file: outputFile ?? `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.${extension}`,
            }),
          paths,
          format,
          sourcemap: true,
          preserveModules,
          exports: 'named',
          globals: {},
          preserveModulesRoot: 'src',
        },
      })),
  )
  .flat();

export default [
  ...[
    outputFileName,
    `${outputFileName}.prefixed`,
    `${outputFileName}.suffixed`,
    'dynamic',
    'build'
  ].map(filename => (
    {
      input: `./src/${filename}.ts`,
      output: [{
        file: `dist/esm/${filename}.d.ts`, format: 'esm'
      }, {
        file: `dist/cjs/${filename}.d.cts`, format: 'cjs'
      }, {
        file: `dist/umd/${filename}.d.ts`, format: 'umd'
      }],
      plugins: [dts()],
    })
  ),
  ...configs,
];
