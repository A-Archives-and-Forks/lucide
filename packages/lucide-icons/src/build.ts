export { default as buildLucideIconNode } from './buildLucideIconNode';
export { default as buildLucideDataUri } from './buildLucideDataUri';
export { default as buildLucideIconElement } from './buildLucideIconElement';
export { default as buildLucideSvg } from './buildLucideSvg';
