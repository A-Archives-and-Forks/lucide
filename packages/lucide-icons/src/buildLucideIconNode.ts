import { LucideBuildParams, LucideIcon, LucideIconNode } from './types';

const defaultAttributes = {
  xmlns: 'http://www.w3.org/2000/svg',
  width: '24',
  height: '24',
  viewBox: '0 0 24 24',
  fill: 'none',
  stroke: 'currentColor',
  'stroke-width': '2',
  'stroke-linecap': 'round',
  'stroke-linejoin': 'round',
};

function buildLucideIconNode(icon: LucideIcon, params: LucideBuildParams = {}): LucideIconNode {
  const viewBoxWidth = ('size' in icon ? icon.size : icon.width) ?? defaultAttributes.width;
  const viewBoxHeight = ('size' in icon ? icon.size : icon.height) ?? defaultAttributes.height;
  const attributes = {
    ...defaultAttributes,
    ...('color' in params && { stroke: params['color'] }),
    ...('size' in params &&
      params['size'] && {
        width: params['size'].toString(10),
        height: params['size'].toString(10),
      }),
    ...('width' in params && params['width'] && { width: params['width'].toString(10) }),
    ...('height' in params && params['height'] && { height: params['height'].toString(10) }),
    ...('strokeWidth' in params &&
      params['strokeWidth'] && { 'stroke-width': params['strokeWidth'].toString(10) }),
    ...('absoluteStrokeWidth' in params &&
      params['absoluteStrokeWidth'] && { 'vector-effect': 'non-scaling-stroke' }),
    class: `lucide lucide-${icon.name} ${params['className'] ?? ''}`.trim(),
    viewBox: `0 0 ${viewBoxWidth} ${viewBoxHeight}`,
    ...('attributes' in params && params.attributes),
  };
  return ['svg', attributes, icon.node];
}

export default buildLucideIconNode;
