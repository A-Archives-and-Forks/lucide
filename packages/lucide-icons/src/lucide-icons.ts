export * as icons from './icons';
export * from './icons';
export * from './aliases/index';
export * from './types';
