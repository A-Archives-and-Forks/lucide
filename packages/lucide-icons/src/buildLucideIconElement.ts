import { LucideBuildParams, LucideIcon, LucideIconNode } from './types';
import buildLucideIconNode from './buildLucideIconNode';

const buildDomElement = (
  document: Document,
  [tagName, attributes, children = []]: LucideIconNode,
): Element => {
  const element = document.createElementNS('http://www.w3.org/2000/svg', tagName);
  for (const [attrName, value] of Object.entries(attributes)) {
    element.setAttribute(attrName, value);
  }
  for (const node of children) {
    const childNode = buildDomElement(document, node);
    element.appendChild(childNode);
  }
  return element;
};

function buildLucideIconElement(
  document: Document,
  icon: LucideIcon,
  params: LucideBuildParams = {},
) {
  return buildDomElement(document, buildLucideIconNode(icon, params));
}

export default buildLucideIconElement;
