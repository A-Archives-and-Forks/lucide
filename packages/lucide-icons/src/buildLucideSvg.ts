import { LucideBuildParams, LucideIcon, LucideIconNode } from './types';
import buildLucideIconNode from './buildLucideIconNode';

const buildDomNode = ([tagName, attributes, children = []]: LucideIconNode): string =>
  `<${tagName} ${Object.entries(attributes)
    .map(([attrName, value]) => `${attrName}="${value}"`)
    .join(' ')}>${children?.map((child) => buildDomNode(child)).join('')}</${tagName}>`;

function buildLucideSvg(icon: LucideIcon, params: LucideBuildParams = {}) {
  return buildDomNode(buildLucideIconNode(icon, params));
}

export default buildLucideSvg;
