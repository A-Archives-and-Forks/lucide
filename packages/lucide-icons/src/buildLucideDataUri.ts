import { LucideBuildParams, LucideIcon } from './types';
import buildLucideSvg from './buildLucideSvg';

function buildLucideDataUri(icon: LucideIcon, params: LucideBuildParams = {}) {
  return `data:image/svg+xml;base64,${btoa(buildLucideSvg(icon, params))}`;
}

export default buildLucideDataUri;
