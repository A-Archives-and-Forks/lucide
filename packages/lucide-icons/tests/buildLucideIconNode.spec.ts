import { describe, expect, it } from 'vitest';
import { House } from '../src/lucide-icons';
import buildLucideIconNode from '../src/buildLucideIconNode';

describe('buildLucideIconNode', () => {
  it('should create icon node', () => {
    const HouseSVG = buildLucideIconNode(House);

    expect(HouseSVG.at(0)).toBe('svg');
  });

  it('should match the snapshot', () => {
    const HouseSVG = buildLucideIconNode(House);

    expect(JSON.stringify(HouseSVG.at(1))).toMatchSnapshot();
  });

  it('should override dimensions, but not viewBox', () => {
    const HouseSVG = buildLucideIconNode(House, { size: 12 });

    expect(HouseSVG[1]['width']).toBe('12');
    expect(HouseSVG[1]['height']).toBe('12');
    expect(HouseSVG[1]['viewBox']).toBe('0 0 24 24');
  });

  it('should override width, but not height', () => {
    const HouseSVG = buildLucideIconNode(House, { width: 12 });

    expect(HouseSVG[1]['width']).toBe('12');
    expect(HouseSVG[1]['height']).toBe('24');
    expect(HouseSVG[1]['viewBox']).toBe('0 0 24 24');
  });

  it('should override height, but not width', () => {
    const HouseSVG = buildLucideIconNode(House, { height: 12 });

    expect(HouseSVG[1]['width']).toBe('24');
    expect(HouseSVG[1]['height']).toBe('12');
    expect(HouseSVG[1]['viewBox']).toBe('0 0 24 24');
  });

  it('should override color', () => {
    const HouseSVG = buildLucideIconNode(House, { color: 'pink' });

    expect(HouseSVG[1]['stroke']).toBe('pink');
    expect(HouseSVG[1]['fill']).toBe('none');
  });

  it('should override stroke width', () => {
    const HouseSVG = buildLucideIconNode(House, { strokeWidth: 12 });

    expect(HouseSVG[1]['stroke-width']).toBe('12');
  });

  it('should set non-scaling-stroke', () => {
    const HouseSVG = buildLucideIconNode(House, { absoluteStrokeWidth: true });

    expect(HouseSVG[1]['vector-effect']).toBe('non-scaling-stroke');
  });

  it('should not set non-scaling-stroke', () => {
    const HouseSVG = buildLucideIconNode(House, { absoluteStrokeWidth: false });

    expect(HouseSVG[1]['vector-effect']).toBeUndefined();
  });

  it('should create icon node with attributes', () => {
    const HouseSVG = buildLucideIconNode(House, { attributes: { fill: 'red' } });

    expect(HouseSVG[1]['fill']).toBe('red');
  });

  it('should create icon node with class name', () => {
    const HouseSVG = buildLucideIconNode(House, { attributes: { class: 'icon' } });

    expect(HouseSVG[1]['class']).toBe('icon');
  });

  it('should merge classes', () => {
    const HouseSVG = buildLucideIconNode(House, { className: 'icon' });

    expect(HouseSVG[1]['class']).toBe('lucide lucide-house icon');
  });
});
