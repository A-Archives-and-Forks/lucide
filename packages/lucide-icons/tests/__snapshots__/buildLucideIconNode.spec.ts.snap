// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildLucideIconNode > should match the snapshot 1`] = `"{"xmlns":"http://www.w3.org/2000/svg","width":"24","height":"24","viewBox":"0 0 24 24","fill":"none","stroke":"currentColor","stroke-width":"2","stroke-linecap":"round","stroke-linejoin":"round","class":"lucide lucide-house"}"`;
